base/dcp_common.go

```diff
@@ -536,3 +540,129 @@ func GenerateDcpStreamName(feedID string) (string, error) {
 	), nil
 
 }
+
+// getExternalAlternateAddress returns a external alternate address for a given dest
+func getExternalAlternateAddress(alternateAddressMap map[string]string, dest string) (string, error) {
+
+	destHost, destPort, err := SplitHostPort(dest)
+	if err != nil {
+		return "", err
+	}
+
+	// Map the given destination to an external alternate address hostname if available
+	if extHostname, foundAltAddress := alternateAddressMap[destHost]; foundAltAddress {
+		_, port, _ := SplitHostPort(extHostname)
+		if port == "" {
+			port = destPort
+		}
+
+		Tracef(KeyDCP, "Found alternate address mapping %s => %s", MD(dest), MD(extHostname+":"+port))
+		dest = extHostname + ":" + port
+	}
+
+	return dest, nil
+}
+
+// alternateAddressShims returns the 3 functions that wrap around ConnectBucket/Connect/ConnectTLS to provide alternate address support.
+func alternateAddressShims(externalAlternateAddresses map[string]string) (
+	connectBucketShim func(serverURL, poolName, bucketName string, auth couchbase.AuthHandler) (cbdatasource.Bucket, error),
+	connectShim func(protocol, dest string) (*memcached.Client, error),
+	connectTLSShim func(protocol, dest string, tlsConfig *tls.Config) (*memcached.Client, error),
+) {
+
+	// Copy of cbdatasource's default ConnectBucket function, which maps internal addresses to alternate addresses
+	connectBucketShim = func(serverURL, poolName, bucketName string, auth couchbase.AuthHandler) (cbdatasource.Bucket, error) {
+		Tracef(KeyDCP, "ConnectBucket callback: %s %s %s", MD(serverURL), poolName, MD(bucketName))
+
+		var (
+			err    error
+			client couchbase.Client
+		)
+
+		if auth != nil {
+			client, err = couchbase.ConnectWithAuth(serverURL, auth)
+		} else {
+			client, err = couchbase.Connect(serverURL)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		pool, err := client.GetPool(poolName)
+		if err != nil {
+			return nil, err
+		}
+
+		// Build map of external alternate addresses if available
+		externalAlternateAddresses = make(map[string]string, len(pool.Nodes))
+		for _, node := range pool.Nodes {
+			if external, ok := node.AlternateNames["external"]; ok && external.Hostname != "" {
+				nodeHostname, _, err := SplitHostPort(node.Hostname)
+				if err != nil {
+					return nil, err
+				}
+
+				// TODO: check if host is tls
```

Comment:
This looks like it's still pending?

Reply:
oops, yeah - this TODO is associated with the `if true {` below - was intending to come back to this. I'll update tomorrow if I get chance.